daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilUnparseContentHandler.scala

```diff
@@ -0,0 +1,309 @@
+   * can just take off the top mappings, because the element that might have cared about the order
+   * is already done using the prefixMappings
+   */
+  override def endPrefixMapping(prefix: String): Unit = {
+    prefixMapping = if (prefixMapping == null) prefixMapping else prefixMapping.parent
+  }
+
+  /**
+   * Uses Attributes, which is passed in to the startElement callback, to extract prefix mappings and
+   * populate the global prefixMapping
+   */
+  def mapPrefixMappingFromAttributesImpl(atts:Attributes): Unit = {
+    var i = 0
+    while (i < atts.getLength) {
+      val qName = atts.getQName(i)
+      if (qName.startsWith("xmlns")) {
+        val uri =  atts.getValue(i)
+        val prefix = if(qName.contains(":")) {
+          val pref = qName.split(":").last
+          pref
+        } else {
+         null // NamespaceBinding does not allow blanks so return null instead
+        }
+        prefixMapping = NamespaceBinding(prefix, uri, prefixMapping)
+      }
+      i += 1
+    }
+  }
+
+  override def startElement(uri: String, localName: String, qName: String, atts: Attributes): Unit = {
+    // we need to check if the characters data is all whitespace, if it is we drop the whitespace
+    // data, if it is not, it is an error as starting a new element with actual characterData means
+    // we haven't hit an endElement yet, which means we're in a complexElement and a complexElement
+    // cannot have character content
+    if (characterData.nonEmpty && !Misc.isAllWhitespace(characterData)) {
+      throw new IllegalContentWhereEventExpected("Non-whitespace characters in complex " +
+        "Element: " + characterData.toString
+      )
+    } else {
+      // reset since it was whitespace only
+      characterData.setLength(0)
+    }
+
+    if (!contentHandlerPrefixMappingUsed) {
+      // always pushes but doesn't always add a mapping since atts can be empty
+      prefixMappingTrackingStack.push(prefixMapping)
+      mapPrefixMappingFromAttributesImpl(atts)
+    }
+
+    if (!infosetEvent.isEmpty && infosetEvent.localName.isDefined) {
+      // we started another element while we were in the process of building a startElement
+      // this means the first element was complex and we are ready for the inputter queue
+      sendToInputter()
+    }
+    // use Attributes to determine xsi:nil value
+    val nilIn = atts.getIndex(XMLConstants.W3C_XML_SCHEMA_INSTANCE_NS_URI, "nil")
+    infosetEvent.nilValue = if (nilIn >= 0) {
+      val nilValue = atts.getValue(nilIn)
+      One(nilValue)
+    } else {
+      Nope
+    }
+    // set localName and namespaceURI
+    lazy val qNameArr = qName.split(":")
+    infosetEvent.localName =
+      if (localName.nonEmpty) {
+        One(localName)
+      } else if (qNameArr.length > 1) {
+        One(qNameArr.last)
+      } else if (qNameArr.nonEmpty) {
+        One(qNameArr.head)
+      } else {
```

Comment:
What is the difference between the `qNameArr.length > 1` and `qNameArr.nonEmpty` cases which requires returning `One(qNameArr.last)` on one hand and `One(qNameArr.head)` on the other hand?  Could you merge both cases into `qNameArr.nonEmpty` and return `One(qNameArr.last)` since `last == head` when `qNameArr.length == 1`?  I suspect there must be a reason since I see the simpler logic used in endElement at lines 232-239 below.

Reply:
Ah you're right, it was a left over from when I was using options, I was returning either lift(1) or lift(0) then since the former case represented when there was a prefix and a localName and the latter, when there is only a localName, but it can be simplified to .nonEmpty and .last